### PR TITLE
add support for setting extraIPv6CIDRs

### PR DESCRIPTION
**Pull request by aclevername**
### Description

Add support for configuring `vpc.extraIPv6CIDRs`

Closes https://github.com/weaveworks/eksctl/issues/4279

## Manual test
config:
```
# An example of ClusterConfig object with custom VPC ipFamily
# set to ipv6; also without any nodegroups:
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: jk-extra-2
  region: us-west-2
  version: "1.21"

vpc:
  ipFamily: IPv6
  extraIPv6CIDRs:
  - 2600:1f13:c3b:5000::/56

addons:
  - name: vpc-cni
  - name: coredns
  - name: kube-proxy

iam:
  withOIDC: true

managedNodeGroups: []
```

```
eksctl create cluster -f examples/29-vpc-with-ip-family.yaml
2021-11-04 12:02:11 [ℹ]  eksctl version 0.70.0-dev+fc554c54.2021-11-04T12:01:13Z
2021-11-04 12:02:11 [ℹ]  using region us-west-2
2021-11-04 12:02:12 [ℹ]  setting availability zones to [us-west-2a us-west-2c us-west-2d]
2021-11-04 12:02:12 [ℹ]  subnets for us-west-2a - public:192.168.0.0/19 private:192.168.96.0/19
2021-11-04 12:02:12 [ℹ]  subnets for us-west-2c - public:192.168.32.0/19 private:192.168.128.0/19
2021-11-04 12:02:12 [ℹ]  subnets for us-west-2d - public:192.168.64.0/19 private:192.168.160.0/19
2021-11-04 12:02:12 [ℹ]  using Kubernetes version 1.21
2021-11-04 12:02:12 [ℹ]  creating EKS cluster "jk-extra-2" in "us-west-2" region with
2021-11-04 12:02:12 [ℹ]  will create a CloudFormation stack for cluster itself and 0 nodegroup stack(s)
2021-11-04 12:02:12 [ℹ]  will create a CloudFormation stack for cluster itself and 0 managed nodegroup stack(s)
2021-11-04 12:02:12 [ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=us-west-2 --cluster=jk-extra-2'
2021-11-04 12:02:12 [ℹ]  CloudWatch logging will not be enabled for cluster "jk-extra-2" in "us-west-2"
2021-11-04 12:02:12 [ℹ]  you can enable it with 'eksctl utils update-cluster-logging --enable-types={SPECIFY-YOUR-LOG-TYPES-HERE (e.g. all)} --region=us-west-2 --cluster=jk-extra-2'
2021-11-04 12:02:12 [ℹ]  Kubernetes API endpoint access will use default of {publicAccess=true, privateAccess=false} for cluster "jk-extra-2" in "us-west-2"
2021-11-04 12:02:12 [ℹ]  3 sequential tasks: { create cluster control plane "jk-extra-2", set AssignIpv6AddressOnCreation to true for public subnets, 5 sequential sub-tasks: { wait for control plane to become ready, associate IAM OIDC provider, no tasks, restart daemonset "kube-system/aws-node", 1 task: { create addons } } }
2021-11-04 12:02:12 [ℹ]  building cluster stack "eksctl-jk-extra-2-cluster"
2021-11-04 12:02:13 [ℹ]  deploying stack "eksctl-jk-extra-2-cluster"
2021-11-04 12:02:43 [ℹ]  waiting for CloudFormation stack "eksctl-jk-extra-2-cluster"
2021-11-04 12:03:14 [ℹ]  waiting for CloudFormation stack "eksctl-jk-extra-2-cluster"
2021-11-04 12:04:15 [ℹ]  waiting for CloudFormation stack "eksctl-jk-extra-2-cluster"
2021-11-04 12:05:16 [ℹ]  waiting for CloudFormation stack "eksctl-jk-extra-2-cluster"
...
```

Checking `ControlPlaneSecurityGroup` security group:
```
Name | Security group rule ID | IP version | Type | Protocol | Port range | Source | Description
-- | -- | -- | -- | -- | -- | -- | --
  | – | sgr-0162d6dec16cde182 | IPv6 | HTTPS | TCP | 443 | 2600:1f13:c3b:5000::/56 | Allow Extra IPv6 CIDR

```
### Checklist
- [X] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [X] Manually tested
- [X] Made sure the title of the PR is a good description that can go into the release notes
- [X] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [X] Backfilled missing tests for code in same general area :tada:
- [X] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: pkg/apis/eksctl.io/v1alpha5/validation_test.go [hunk 1]
```diff
@@ -708,6 +708,21 @@ var _ = Describe("ClusterConfig validation", func() {
...
 		})
 
+		Context("extraIPv6CIDRs", func() {
+			It("validates cirds", func() {
+				cfg.VPC.ExtraIPv6CIDRs = []string{"2002::1234:abcd:ffff:c0a8:101/64"}
+				err = cfg.ValidateVPCConfig()
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			When("extraCIDRs has an invalid cidr", func() {
```
Comment (Skarlso) [inline pkg/apis/eksctl.io/v1alpha5/validation_test.go:718]: can you add one please with just an invalid block identifier?

Comment (aclevername) [reply pkg/apis/eksctl.io/v1alpha5/validation_test.go:718]: done

File: pkg/cfn/builder/cluster_test.go [hunk 2]
```diff
@@ -172,34 +172,65 @@ var _ = Describe("Cluster Template Builder", func() {
...
+				Expect(clusterTemplate.Resources).To(HaveKey("IngressControlPlaneExtraCIDR1"))
+				Expect(clusterTemplate.Resources["IngressControlPlaneExtraCIDR1"].Properties).To(Equal(fakes.Properties{
+					CidrIP:     "192.168.1.0/24",
+					IPProtocol: "tcp",
+					FromPort:   443,
+					ToPort:     443,
+					GroupID: map[string]interface{}{
+						"Ref": "ControlPlaneSecurityGroup"},
+					Description: "Allow Extra CIDR 1 (192.168.1.0/24) to communicate to controlplane",
+				}))
```
Comment (Skarlso) [inline pkg/cfn/builder/cluster_test.go:202]: ? Just a precaution that multiple rules can be defined?

**Review by Skarlso — APPROVED** [review]
Nothing unexpected. Noice!

File: pkg/cfn/builder/cluster_test.go (hunk 2 above)
Comment (aclevername) [reply pkg/cfn/builder/cluster_test.go:202]: got to have a test for the for-loop :wink:

File: pkg/apis/eksctl.io/v1alpha5/validation.go [hunk 3]
```diff
@@ -161,6 +161,13 @@ func (c *ClusterConfig) ValidateVPCConfig() error {
 		}
 		c.VPC.PublicAccessCIDRs = cidrs
 	}
+	if len(c.VPC.ExtraIPv6CIDRs) > 0 {
```
Comment (nikimanoledaki) [inline pkg/apis/eksctl.io/v1alpha5/validation.go:164]: a user should only be able to set `vpc.extraIPv6CIDRs` if they've also set `vpc.IpFamily` to `IPv6`

File: pkg/apis/eksctl.io/v1alpha5/validation_test.go [hunk 4]
```diff
@@ -685,7 +685,7 @@ var _ = Describe("ClusterConfig validation", func() {
 			})
 		})
 
-		Context("CIDRs", func() {
+		Context("extraCIDRs", func() {
 			It("validates cirds", func() {
```
Comment (nikimanoledaki) [inline pkg/apis/eksctl.io/v1alpha5/validation_test.go:689]:
```suggestion
			It("validates cidrs", func() {
```
soz 😅

File: pkg/apis/eksctl.io/v1alpha5/validation_test.go [hunk 5]
```diff
@@ -708,6 +708,25 @@ var _ = Describe("ClusterConfig validation", func() {
 			})
 		})
 
+		Context("extraIPv6CIDRs", func() {
+			It("validates cirds", func() {
```
Comment (nikimanoledaki) [inline pkg/apis/eksctl.io/v1alpha5/validation_test.go:712]:
```suggestion
			It("validates cidrs", func() {
```

File: pkg/apis/eksctl.io/v1alpha5/validation_test.go [hunk 6]
```diff
@@ -708,6 +708,25 @@ var _ = Describe("ClusterConfig validation", func() {
...
+				cfg.VPC.ExtraIPv6CIDRs = []string{"2002::1234:abcd:ffff:c0a8:101/64"}
+				err = cfg.ValidateVPCConfig()
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			When("extraCIDRs has an invalid cidr", func() {
```
Comment (nikimanoledaki) [inline pkg/apis/eksctl.io/v1alpha5/validation_test.go:718]:
```suggestion
			When("extraIPv6CIDRs has an invalid cidr", func() {
```

File: pkg/cfn/builder/cluster_test.go [hunk 7]
```diff
@@ -172,34 +172,65 @@ var _ = Describe("Cluster Template Builder", func() {
...
 
-			It("should add 1 extra control plane ingress rule", func() {
+			It("should add extra control plane ingress rules", func() {
 				Expect(clusterTemplate.Resources).To(HaveKey("IngressControlPlaneExtraCIDR0"))
-				Expect(clusterTemplate.Resources["IngressControlPlaneExtraCIDR0"].Properties.CidrIP).To(Equal("192.168.0.0/24"))
-				Expect(clusterTemplate.Resources["IngressControlPlaneExtraCIDR0"].Properties.IPProtocol).To(Equal("tcp"))
-				Expect(clusterTemplate.Resources["IngressControlPlaneExtraCIDR0"].Properties.FromPort).To(Equal(443))
-				Expect(clusterTemplate.Resources["IngressControlPlaneExtraCIDR0"].Properties.ToPort).To(Equal(443))
+
+				Expect(clusterTemplate.Resources["IngressControlPlaneExtraCIDR0"].Properties).To(Equal(fakes.Properties{
```
Comment (nikimanoledaki) [inline pkg/cfn/builder/cluster_test.go:183]: yay for backfilling/improving past tests!!! 🎉 💯

File: userdocs/src/usage/vpc-networking.md [hunk 8]
```diff
@@ -456,7 +456,8 @@ There are some additional caveats when configuring Kubernetes API endpoint acces
    resources.  See:
    [EKS user guide](https://docs.aws.amazon.com/en_pv/eks/latest/userguide/cluster-endpoint#private-access)
    A user can elect to supply vpc.extraCIDRs which will append additional CIDR ranges to the ControlPlaneSecurityGroup,
-   allowing subnets outside the VPC to reach the kubernetes API endpoint.
+   allowing subnets outside the VPC to reach the kubernetes API endpoint. Similarly you can provide `vpc.extraIPv6CIDR`
```
Comment (nikimanoledaki) [inline userdocs/src/usage/vpc-networking.md:459]:
```suggestion
   allowing subnets outside the VPC to reach the kubernetes API endpoint. Similarly you can provide `vpc. ExtraIPv6CIDRs`
```

**Review by nikimanoledaki — CHANGES_REQUESTED** [review]
awesome stuff! 😄

File: pkg/apis/eksctl.io/v1alpha5/validation.go (hunk 3 above)
Comment (aclevername) [reply pkg/apis/eksctl.io/v1alpha5/validation.go:164]: eeeek good spot!

File: userdocs/src/usage/vpc-networking.md (hunk 8 above)
Comment (aclevername) [reply userdocs/src/usage/vpc-networking.md:459]: that adds a space :thinking:

Comment (nikimanoledaki) [reply userdocs/src/usage/vpc-networking.md:459]: oops - edited it!

**Comment by Skarlso** [comment]
Wow @nikimanoledaki frigging awesome job on this review. I have not spotted any of those. WELL DONE! 🎉

File: pkg/apis/eksctl.io/v1alpha5/validation.go (hunk 3 above)
Comment (aclevername) [reply pkg/apis/eksctl.io/v1alpha5/validation.go:164]: done!

**Review by nikimanoledaki — APPROVED** [review]
one last thing then lgtm! 🚀 

edit: also thank you @Skarlso, aw !! 💛

**Pull request merged**